### ISSUE TITLE
Ability to add patches alongside the formula

### ISSUE THREAD

**Issue by gaborbernat** [Homebrew/brew#15838]
### Verification

- [X] This issue's title and/or description do not reference a single formula e.g. `brew install wget`. If they do, open an issue at https://github.com/Homebrew/homebrew-core/issues/new/choose instead.

### Provide a detailed description of the proposed feature

Currently you can add patches as external URI or embeded. However, I'd like to add patches alongside the formula, within the tap.

### What is the motivation for the feature?

This way I can apply formatter/linter to the formula files without conflicting with the patch logic (e.g. stripping trailing whitespace).

### How will the feature be relevant to at least 90% of Homebrew users?

Will make maintaining brew formulas much easier.

### What alternatives to the feature have been considered?

Didn't see any alternatives.

**Comment by MikeMcQuaid** [comment]
If the patch is in the tap: you can use the URL to the patch in the tap in the formula itself. Passing on this but will review a PR if you wanted to create one.

**Comment by gaborbernat** [comment]
Yeah but now I'm hardcoding my hosted Github Enterprise URI in the file, not to mention this might run fault of authentication requirements (which the ssh key clonse authentication needed to use the tap doesn't cover).  Ideally I'd like to do instead something like:

```rb
patch do
  files a_b_c_d.diff a_b_c_e.diff
end
```

**Comment by gaborbernat** [comment]
Would that work for you?

**Comment by gaborbernat** [comment]
I wish you would have kept the issue at least until we reach a consensus on this topic and not close it right away...

**Comment by MikeMcQuaid** [comment]
> ```ruby
> patch do
>   files a_b_c_d.diff a_b_c_e.diff
> end
> ```
>
> Would that work for you?

This would work:

```ruby
patch do
  tap_file "a_b_c_d.diff"
end
```

> I wish you would have kept the issue at least until we reach a consensus on this topic and not close it right away...

That's not how we work this issue tracker, sorry. You're the first person I've heard requesting this so it's firmly in the territory of "we'd review a PR and merge a good one but aren't actively soliciting one and no maintainer is going to pick this up".

We can continue to discuss on a closed issue.

**Comment by gaborbernat** [comment]
I have never written Ruby before, do you have any guide how to install deps and run the test suite? Some documentation on this?

**Comment by scpeters** [comment]
> I have never written Ruby before, do you have any guide how to install deps and run the test suite? Some documentation on this?

you should be able to run the [patch_spec.rb](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/test/patch_spec.rb) test with the following command:

~~~
brew tests --only=patch
~~~

this automatically downloads dependencies on my system

all the brew documentation is at docs.brew.sh, though it's mostly geared towards users and homebrew-core contributors. There is a [guide for maintainers of homebrew/brew](https://docs.brew.sh/Homebrew-brew-Maintainer-Guide), but it doesn't include an introduction to ruby.
